keps/sig-network/20190415-Autopath API for clusterDNS.md

```diff
@@ -0,0 +1,104 @@
+Even if one of the several queries fails due, the DNS lookup on the client side will fail after a 5s timeout.
+
+### Goals
+Provide a solution to minimize number of DNS queries on the client side, for a DNS lookup. 
+This solution will preserve the current behavior that allows shortname lookups of Kubernetes resources. Make this solution configurable via the API.
+
+### Non-Goals
+Modifying the current Kubernetes DNS Schema is not a goal of this KEP.
+
+
+## Proposal
+
+This proposal introduces a new dnsPolicy "clusterFirstWithAutopath". This can be set by including the line below in podSpec:
+
+ ` dnsPolicy: clusterFirstWithAutopath`
+
+Using this mode will set the searchpath on the pod as `search.$NS.$SUFFIX.k`, where $NS is the namespace of the pod and $SUFFIX is the cluster suffix. k is a one-letter suffix to identify the domain name to be a query that needs search expansion. Since there are [no single-letter TLD so far](http://data.iana.org/TLD/tlds-alpha-by-domain.txt), this suffix will help identify queries needing search expansion. 
+
+This approach minimizes the number of DNS queries at client side to atmost 2(A, AAAA). The searchpath expansion logic moves to the server side. The server(clusterDNS - CoreDNS by default) will require additional metadata in order to complete searchpath expansion.
+
+The metadata can be attached as an EDNS0 option. We can define a new EDNS0 option - SearchPaths, option code: 15 that is currently [unassigned.](https://www.iana.org/assignments/dns-parameters/dns-parameters.xhtml#dns-parameters-11) We would need to reserve this or use an option number from the experimental range - 65001 to 65534.
+The value of this option will be a comma-separated string consisting of all the searchpaths which are to be appended to the main query and looked up. This option can be useful outside of Kubernetes as well.
+
+Instead of modifying all client pod images to insert an EDNS0 option in their requests, we will put this logic in [Nodelocal DNSCache](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/0030-nodelocal-dns-cache.md), which is a daemonset running a per-node DNS Cache.
+All pods using clusterDNS will send requests to the local instance running on the same node. This instance will look for the custom suffix that we defined above, in incoming query names. In case of a match, it will extract the namespace and cluster suffix values and construct the EDNS0 option with the searchpath strings.
```

Comment:
This defeats the HA discussion for node-local cache, doesn't it?  You can't flip back to cluster DNS -- it won't know the magic.

But if we push the magic to the server, then it needs to know the expansion schema.  Yuck

Reply:
Yes, the current HA plan proposed in https://github.com/kubernetes/enhancements/pull/1005 won't work if the searchpath is set to something only node-local cache can understand. I added this to the "risks" as well.
Pushing it to server, specifically CoreDNS might not be that bad.. the current autopath plugin in CoreDNS supports getting searchpaths from another plugin. kubernetes plugin has an implementation to provide the list of paths today - https://github.com/coredns/coredns/blob/72a633ba0956f771342f1f2993696f37e1dee6db/plugin/kubernetes/autopath.go#L11
We could extend this to match the domain name suffix and figure it out. Maybe pass in a version too.

Reply:
As long as the magic string unambiguously contains the client pod's namespace, putting the logic in the server is not an issue. It would just be a change to the current autopath feature. The magic string would have all of the context so there would be no need for the pod watch.

The only real value of putting this into an EDNS0 option is that: 1) the EDNS0-based search path is usable in other contexts; 2) the magic domain name is tightly coupled to this particular use case rather than being more broadly visible; and 3) we could expand the EDNS0 option to include other info and implement more complex server-side DNS policy (needs use cases...).

Reply:
I agree with putting the logic in the server to parse out the required fields and generate the searchpaths. The only concern is that then this feature is supported only when CoreDNS is the DNS server. That might be ok to start with since it is the default clusterDNS server now.